src/workerd/api/headers.c++

```diff
@@ -1,134 +1,367 @@
+  JSG_REQUIRE(workerd::util::isValidHeaderValue(trimmed), TypeError, "Invalid header value.");
+  maybeWarnIfBadHeaderString(value);
+  // By attaching the original array to the trimmed view, we keep the original allocation alive
+  // and prevent an unnecessary copy.
+  return kj::str(trimmed.attach(value.releaseArray()));
+}
 
-  constexpr auto HTTP_TOKEN_CHARS = kj::parse::controlChar.orChar('\x7f')
-                                        .orGroup(kj::parse::whitespaceChar)
-                                        .orGroup(HTTP_SEPARATOR_CHARS)
-                                        .invert();
-  // RFC2616 section 2.2: https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.2
-  // RFC2616 section 4.2: https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+Headers::HeaderKey getHeaderKeyFor(kj::StringPtr name) {
+  if (uint commonId = HEADER_HASH_TABLE.find(name)) {
+    KJ_DASSERT(commonId > 0 && commonId <= Headers::MAX_COMMON_HEADER_ID);
+    return commonId;
+  }
 
   for (char c: name) {
-    JSG_REQUIRE(HTTP_TOKEN_CHARS.contains(c), TypeError, "Invalid header name.");
+    JSG_REQUIRE(util::isHttpTokenChar(c), TypeError, "Invalid header name.");
   }
-}
 
-void requireValidHeaderValue(kj::StringPtr value) {
-  // TODO(cleanup): Code duplication with kj/compat/http.c++
-
-  for (char c: value) {
-    JSG_REQUIRE(c != '\0' && c != '\r' && c != '\n', TypeError, "Invalid header value.");
-  }
+  // Not a common header, so allocate lowercase copy for uncommon header
+  return toLower(name);
 }
 }  // namespace
 
-Headers::Headers(jsg::Lock& js, jsg::Dict<jsg::ByteString, jsg::ByteString> dict)
-    : guard(Guard::NONE) {
+Headers::Headers(jsg::Lock& js, jsg::Dict<kj::String, kj::String> dict): guard(Guard::NONE) {
+  // Because the headers might end up in either of our two tables,
+  // we can't really reserve space for them up front.
   for (auto& field: dict.fields) {
     append(js, kj::mv(field.name), kj::mv(field.value));
   }
 }
 
 Headers::Headers(jsg::Lock& js, const Headers& other): guard(Guard::NONE) {
-  for (auto& header: other.headers) {
-    Header copy{
-      jsg::ByteString(kj::str(header.second.key)),
-      jsg::ByteString(kj::str(header.second.name)),
-      KJ_MAP(value, header.second.values) { return jsg::ByteString(kj::str(value)); },
-    };
-    kj::StringPtr keyRef = copy.key;
-    KJ_ASSERT(headers.insert(std::make_pair(keyRef, kj::mv(copy))).second);
+  for (kj::uint i = 0; i < other.commonHeaders.size(); i++) {
+    commonHeaders[i] =
+        other.commonHeaders[i].map([](const kj::Own<Header>& h) { return h->clone(); });
+  }
+  uncommonHeaders.reserve(other.uncommonHeaders.size());
+  for (auto& [key, header]: other.uncommonHeaders) {
+    // It should not be possible to have duplicate keys here.
+    uncommonHeaders.insert(kj::str(key), header->clone());
   }
 }
 
 Headers::Headers(jsg::Lock& js, const kj::HttpHeaders& other, Guard guard): guard(Guard::NONE) {
```

Comment:
You could initialize `guard(guard)` directly now, since you use `appendUnguarded()`, I think.